include/nbl/builtin/hlsl/math/functions.hlsl

```diff
@@ -0,0 +1,609 @@
+{
+    return impl::bitfieldInsert(base,value,int(offset),int(count));
+}
+
+
+uint bitfieldInsert(in uint base, uint value, in uint offset, in uint count)
+{
+    const uint shifted_masked_value = (value&((0x1u<<count)-1u))<<offset;
+    return impl::bitfieldInsert(base,shifted_masked_value,base&((0x1u<<offset)-1u),count);
+}
+
+//! Trig
+
+// returns `acos(acos(A)+acos(B)+acos(C))-PI` but requires `sinA,sinB,sinC` are all positive
+float getArccosSumofABC_minus_PI(in float cosA, in float cosB, in float cosC, in float sinA, in float sinB, in float sinC)
+{
+    // sorry about the naming of `something` I just can't seem to be able to give good name to the variables that is consistent with semantics
+   const bool something0 = cosA<(-cosB);
+    const float cosSumAB = cosA*cosB-sinA*sinB;
+   const bool something1 = cosSumAB<(-cosC);
+   const bool something2 = cosSumAB<cosC;
+   // apply triple angle formula
+   const float absArccosSumABC = acos(clamp(cosSumAB*cosC-(cosA*sinB+sinA*cosB)*sinC,-1.f,1.f));
+   return ((something0 ? something2:something1) ? (-absArccosSumABC):absArccosSumABC)+(something0||something1 ? PI:(-PI));
+}
+
+float2 combineCosForSumOfAcos(in float2 cosA, in float2 cosB) 
+{
+   const float bias = cosA.y + cosB.y;
+   const float a = cosA.x;
+   const float b = cosB.x;
+   const bool reverse = abs(min(a,b)) > max(a,b);
+   const float c = a*b - sqrt((1.0f-a*a)*(1.0f-b*b));
+   return reverse ? float2(-c, bias + PI) : float2(c, bias);
+}
+
+// returns acos(a) + acos(b)
+float getSumofArccosAB(in float cosA, in float cosB) 
+{
+   const float2 combinedCos = combineCosForSumOfAcos(float2(cosA, 0.0f), float2(cosB, 0.0f));
+   return acos(combinedCos.x) + combinedCos.y;
+}
+
+// returns acos(a) + acos(b) + acos(c) + acos(d)
+float getSumofArccosABCD(in float cosA, in float cosB, in float cosC, in float cosD) 
+{
+   const float2 combinedCos0 = combineCosForSumOfAcos(float2(cosA, 0.0f), float2(cosB, 0.0f));
+   const float2 combinedCos1 = combineCosForSumOfAcos(float2(cosC, 0.0f), float2(cosD, 0.0f));
+   const float2 combinedCos = combineCosForSumOfAcos(combinedCos0, combinedCos1);
+   return acos(combinedCos.x) + combinedCos.y;
+}
+
+//! MVC
+
+
+// return dFdR (TODO: a completely separate include for this)
+float applyChainRule1D(in float dFdG, in float dGdR)
+{
+   return dFdG*dGdR;
+}
+float2 applyChainRule1D(in float2 dFdG, in float dGdR)
+{
+   return dFdG*dGdR;
+}
+float3 applyChainRule1D(in float3 dFdG, in float dGdR)
+{
+   return dFdG*dGdR;
+}
+float4 applyChainRule1D(in float4 dFdG, in float dGdR)
+{
+   return dFdG*dGdR;
+}
+/* TODO*/
+float1x2 applyChainRule1D(in float dFdG, in float1x2 dGdR)
+{
+   return mul(dFdG, dGdR);
+}
+float2x2 applyChainRule1D(in float2x1 dFdG, in float1x2 dGdR)
+{
+   return mul(dFdG, dGdR);
+}
+float3x2 applyChainRule1D(in float3x1 dFdG, in float1x2 dGdR)
+{
+   return mul(dFdG, dGdR);
+}
+float4x2 applyChainRule1D(in float4x1 dFdG, in float1x2 dGdR)
+{
+   return mul(dFdG, dGdR);
+}
```

Comment:
actually its doable in a single template with https://github.com/microsoft/hlsl-specs/issues/24#issuecomment-1370297244